internal/arena/unsafe.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package arena
+
+import (
+	"runtime"
+	"unsafe"
+)
+
+// pointerIndex returns an integer n such that p == &s[n], or -1 if there is
+// no such integer.
+func pointerIndex[T any](p *T, s []T) int {
+	a := unsafe.Pointer(p)
+	b := unsafe.Pointer(unsafe.SliceData(s))
+	// KeepAlive escapes its argument, so this ensures that a and b have
+	// escaped to the heap and won't be moved.
+	runtime.KeepAlive([2]unsafe.Pointer{a, b})
```

Comment:
The Go docs say to _only_ use `runtime.KeepAlive` to prevent a finalizer from running. So this feels like it could be brittle and thus dangerous.

Also, this construction -- passing an array of pointers instead of a pointer -- is _very_ counter-intuitive. I guess this relies on the fact that pointers in the heap can never point to address on the stack? It feels safe to do this earlier with the original pointers instead of with the `unsafe.Pointer` aliases, something like so:
```go
runtime.KeepAlive(p)
runtime.KeepAlive(&s[0])
```

Reply:
Oh the reason for doing it this way is to only have to pay for one function call. KeepAlive cannot be inlined. Sorry, I am very used to working around Go not having tuples by using arrays. XD

Also, `&s[0]` is incorrect, because it will trigger a redundant bounds check. You still want `unsafe.SliceHeader`.

Separately, reproduced from Slack:

> it's not the right way to do it. there are a few alternatives. one is to open-code KeepAlive, because the current implementation will escape the pointers. we could instead mark the function as go:nosplit. this is used within the runtime in many places as "do not move the pointers I just received as arguments", sufficiently that people have copied the pattern and changing it would almost certainly break the world. note that even if we don't annotate it, the current version of Go won't fuck us: stack growth only happens on function entry, and that is the only situation under which go will move pointers
> also, i struggle to think of a way they could move the pointers out from under us after casting to uintptr, and somehow discriminating from the case where we have an unsafe.Pointer. and also, this requires the move to happen after we've cast one of the pointers but not the others, and the result is that we report that the arena does not contain a pointer it actually contains
> like with my native compiler person hat on i think the risk is nonexistent
> i'm happy to write an essay about this in the comments

Reply:
Yeah, I'm trying to think if we even need anything. Maybe `//go:nosplit` out of an abundance of caution and omit the `runtime.KeepAlive`? The only danger is when a pointer is in a `uintptr` var, in which case it won't be updated if it points to a stack address that gets moved. So the only risk is if stack motion occurred in the middle of the calculation on line 31.

Reply:
Sure let's do that.